### PR TITLE
Require a definition in the same file as an `impl` declaration

### PR DESCRIPTION
**Pull request by josh11b**
This PR detects the failures that #4709 fixes.

### CODE REVIEW

File: toolchain/check/check_unit.cpp [hunk 1]
```diff
@@ -395,8 +395,11 @@ auto CheckUnit::CheckRequiredDefinitions() -> void {
         break;
       }
       case CARBON_KIND(SemIR::ImplDecl impl_decl): {
-        if (!context_.impls().Get(impl_decl.impl_id).is_defined()) {
-          emitter_.Emit(decl_inst_id, MissingDefinitionInImpl);
+        auto& impl = context_.impls().Get(impl_decl.impl_id);
+        if (!impl.is_defined()) {
+          CARBON_DIAGNOSTIC(MissingImplDefinition, Error,
+                            "no definition found for declaration of impl");
```
Comment (zygoloid) [inline toolchain/check/check_unit.cpp]: Maybe "impl declared but not defined"? I think we have previously said that we prefer diagnostics that state facts about the program rather than actions of the toolchain (eg, say the thing doesn't exist, not that we couldn't find it).

File: toolchain/check/impl.cpp [hunk 2]
```diff
@@ -134,27 +134,51 @@ static auto CheckAssociatedFunctionImplementation(
...
-  const auto& interface = context.interfaces().Get(interface_type.interface_id);
+auto ImplWitnessForDeclaration(Context& context, const SemIR::Impl& impl)
+    -> SemIR::InstId {
+  auto facet_type_id = context.GetTypeIdForTypeInst(impl.constraint_id);
+  if (facet_type_id == SemIR::ErrorInst::SingletonTypeId) {
+    return SemIR::ErrorInst::SingletonInstId;
+  }
+  auto facet_type = context.types().TryGetAs<SemIR::FacetType>(facet_type_id);
+  if (!facet_type) {
+    CARBON_DIAGNOSTIC(ImplAsNonFacetType, Error, "impl as non-facet-type");
```
Comment (zygoloid) [inline toolchain/check/impl.cpp]:
```suggestion
    CARBON_DIAGNOSTIC(ImplAsNonFacetType, Error, "impl as non-facet type");
```
Do you think it'd be useful to include the type `facet_type_id` in this diagnostic?

**Review by zygoloid — APPROVED** [review]
LG, assuming the errors go away as expected when the other PR is merged.

**Comment by josh11b** [comment]
I've synced and the four failing tests have been updated:
* check/testdata/impl/declaration.carbon,
* check/testdata/impl/redeclaration.carbon,
* check/testdata/impl/no_prelude/import_generic.carbon,
* check/testdata/impl/no_prelude/generic_redeclaration.carbon

I'll look at your suggestions tomorrow.

File: toolchain/check/impl.cpp (hunk 2 above)
Comment (josh11b) [reply toolchain/check/impl.cpp]: Used `impl.constraint_id` instead of `facet_type_id` so we can eventually use the source's spelling of that type.

File: toolchain/check/check_unit.cpp (hunk 1 above)
Comment (josh11b) [reply toolchain/check/check_unit.cpp]: Done.

File: toolchain/check/handle_impl.cpp [hunk 3]
```diff
@@ -313,17 +326,22 @@ static auto BuildImplDecl(Context& context, Parse::AnyImplDeclId node_id,
...
-      impl_decl.impl_id = prev_impl_id;
+      valid_redeclaration = IsValidImplRedecl(context, impl_info, prev_impl_id);
+      if (valid_redeclaration) {
+        impl_decl.impl_id = prev_impl_id;
+      }
       break;
     }
   }
 
   // Create a new impl if this isn't a valid redeclaration.
```
Comment (danakj) [inline toolchain/check/handle_impl.cpp:341]: FWIW I am finding it hard to follow the logic in this change. Specifically:
- There's now `valid_redeclaration` as a bool.
- The comment here shows we check if we set the `impl_decl.impl_id` to something valid to see if it's a valid redeclaration.
- The bool can be true without ever going into the point where we set `impl_decl.impl_id`. So the bool means something similar, but different.
- Then later we check the bool instead of `impl_decl.impl_id` to see if it's a "valid redeclaration".

Comment (josh11b) [reply toolchain/check/handle_impl.cpp:341]: You are right, this was the result of iterating until it behaved the way I wanted. The purpose of this bool was to plumb the information about whether we've already issued a particular error to the point where we can avoid producing another error. I think it would make sense to rename this `!invalid_redeclaration`, but I don't know if that goes far enough. Does #4738 help?

Comment (danakj) [reply toolchain/check/handle_impl.cpp:341]: yes it helps a lot with the intent of the bool.

I do think there's still a bit of collision with "valid redeclaration" meaning that `impl_decl.impl_id.is_valid()`. Maybe `invalid_redeclaration` would be better using a different name than "valid" and IsValidImplRedecl() could be renamed to match. I am not sure but maybe "legal" or something for that, since it's checking rules, whereas "valid" is used to refer to the id value.

Edit: clarifying grammar
